subprojects/dependency-management/src/main/java/org/gradle/api/internal/attributes/DefaultAttributesSchema.java

```diff
@@ -104,18 +91,17 @@ public boolean hasAttribute(Attribute<?> key) {
         return strategies.containsKey(key);
     }
 
-    AttributeSelectionSchema mergeWith(AttributesSchemaInternal producerSchema) {
-        return new MergedSchema(producerSchema);
-    }
+    private final Map<DefaultAttributeSelectionSchema.ExtraAttributesEntry, Attribute<?>[]> extraAttributesCache = new HashMap<>();
 
     @Override
     public AttributeMatcher withProducer(AttributesSchemaInternal producerSchema) {
-        return new DefaultAttributeMatcher(componentAttributeMatcher, mergeWith(producerSchema));
+        return matcherCache.computeIfAbsent(producerSchema, key ->
+            new ComponentAttributeMatcher(new DefaultAttributeSelectionSchema(this, producerSchema, extraAttributesCache)));
```

Comment:
What's being cached here? It seems like all the instances are going to share the extraAttributesCache so why wouldn't we just create a new one every time?

Reply:
The code you looked at was not the final version. This PR was still a draft. These instances should not be sharing a common cache as their behavior which calculates the cached values may differ.